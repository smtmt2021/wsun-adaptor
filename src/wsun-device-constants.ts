/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export const MAX_BUF_SIZE = 1024;
export const MAX_DGRAM_SIZE = 1024;

const isTesting = process.env.NODE_ENV === 'test';
export const COMMAND_TIMEOUT = (isTesting ? 1 : 3) * 1000; // msec
export const SEND_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const TERM_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const SCAN_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
export const CONNECTION_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
