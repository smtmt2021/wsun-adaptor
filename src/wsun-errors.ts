/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export class WsunError extends Error {
  constructor(message: string) {
    super('wsun: ' + message);
    this.name = new.target.name;
  }
}

export class WsunUdpTransmissionFailed extends WsunError {
  constructor() {
    super("wsun adaptor replied 'udp transmission failed'");
    this.name = new.target.name;
  }
}

export class WsunSendingInProgress extends WsunError {
  constructor() {
    super('wsun is sending in progress');
    this.name = new.target.name;
  }
}

export class WsunCommandError extends WsunError {
  private static reason2message(reason: number): string {
    switch (reason) {
      case 4:
        return "wsun adaptor replied 'unsupported command'";
      case 5:
        return "wsun adaptor replied 'incorrect command'";
      case 6:
        return "wsun adaptor replied 'incorrect range'";
      case 9:
        return "wsun adaptor replied 'uart error'";
      case 10:
        return "wsun adaptor replied 'failed to get result'";
      default:
        return "wsun adaptor replied 'unknown error'";
    }
  }

  constructor(public reason: number) {
    super(WsunCommandError.reason2message(reason));
    this.name = new.target.name;
  }
}

export class WsunTimeoutError extends WsunError {
  constructor(message: string) {
    super(message);
    this.name = new.target.name;
  }
}
