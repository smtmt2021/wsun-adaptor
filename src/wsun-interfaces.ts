/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { RemoteInfo } from 'dgram';

export type WsunEventUdp = 'rxudp';

export interface WsunCache {
  channel: number;
  panId: number;
  addr: string;
}

export interface WsunPanDesc extends WsunCache {
  page?: number;
  pairId?: string;
  LQI?: number;
}

export interface WsunId {
  id: string;
  password: string;
}

export interface WsunUDP {
  sender: string;
  dest: string;
  rport: number;
  lport: number;
  senderlla: string;
  secured: boolean;
  datalen: number;
  data: Buffer;
}

export interface WsunSocket {
  bind(port: number): Promise<void>;
  close(): Promise<void>;
  send(msg: Buffer | string, port: number, address: string): Promise<void>;
  on(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  once(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeAllListeners(event: 'message'): this;
}

export interface WsunConfig {
  device: string; // device name of serial port which is connected with wsun adaptor
}

export interface WsunAdaptorInternal {
  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Delete a used socket.
   *  This function is called automatically whenever a socket is closed.
   *  @param  socket  to be delete.
   */
  deleteSocket(socket: WsunSocket): void;

  /**
   *  Send a datagram message to the specified destination.
   *  This function is called via socket, do not call directly.
   *  @param  msg  to be sent.
   *  @param  port  is destination port.
   *  @param  address is destination address.
   */
  sendDgram(msg: Buffer, port: number, address: string): Promise<void>;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'rxudp'
   */
  on(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;

  /**
   * removes the specified listener from the event
   * @param en
   * @param listener
   */
  removeListener(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;
}
