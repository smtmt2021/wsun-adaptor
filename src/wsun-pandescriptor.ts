/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunPanDesc } from './';

export class PanDescriptor {
  private descriptor = {} as WsunPanDesc;

  reset(): void {
    this.descriptor = {} as WsunPanDesc;
  }

  get isValid(): boolean {
    return !(this.descriptor.channel == null
      || this.descriptor.page == null
      || this.descriptor.panId == null
      || this.descriptor.addr == null
      || this.descriptor.LQI == null
      || this.descriptor.pairId == null
    );
  }

  set channel(arg: string) {
    const channel = parseInt(arg, 16);
    if (!isNaN(channel)) this.descriptor.channel = channel;
  }

  set page(arg: string) {
    const page = parseInt(arg, 16);
    if (!isNaN(page)) this.descriptor.page = page;
  }

  set panId(arg: string) {
    const panId = parseInt(arg, 16);
    if (!isNaN(panId)) this.descriptor.panId = panId;
  }

  set addr(arg: string) {
    this.descriptor.addr = arg;
  }

  set LQI(arg: string) {
    const LQI = parseInt(arg, 16);
    if (!isNaN(LQI)) this.descriptor.LQI = LQI;
  }

  set pairId(arg: string) {
    this.descriptor.pairId = arg;
  }

  getDescriptor(): WsunPanDesc {
    const descriptor = Object.assign({}, this.descriptor);
    this.reset();
    return descriptor;
  }

  toString(): string {
    return JSON.stringify(this.descriptor);
  }
}
