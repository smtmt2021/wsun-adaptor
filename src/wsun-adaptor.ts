/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunId, WsunPanDesc, WsunSocket } from './';
import { WsunCache } from './wsun-interfaces';

export type WsunEventSession =
  | 'session-established'
  | 'session-ending'
  | 'session-ended'
  | 'session-error'
  | 'session-timeout'
  | 'session-108-start'
  | 'session-108-end';

export interface WsunAdaptor {
  /**
   *  Start activities with a wsun adaptor
   *  and establish a connection with PAA coordinator.
   *  @param  wsunId  ID & password
   *  @param  cache   cache info
   *  @returns  a descriptor of the connected PAA coordinator,
   *            undefined if no PAA coordinator was found.
   */
  open(wsunId: WsunId, cache?: WsunCache): Promise<WsunPanDesc | undefined>;

  /**
   *  Terminate a connection with a PAA coordinator
   */
  close(): Promise<void>;

  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Create a socket to communicate remote node.
   *  @return a socket.
   */
  createSocket(): WsunSocket;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'session-established' a PANA session is established
   * @event 'session-ending'      receive a PANA session end request
   * @event 'session-ended'       a PANA session is ended
   * @event 'session-error'       a PANA session is ended due to error
   * @event 'session-timeout'     a PANA session is ended due to timeout
   * @event 'session-108-start'   Restriction of ARIB 108 is started
   * @event 'session-108-end'     Restriction of ARIB 108 is ended
   */
  on(en: WsunEventSession, listener: () => void): void;
}
