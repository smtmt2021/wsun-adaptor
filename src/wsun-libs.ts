/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunCache } from './wsun-interfaces';

export const isParamOfPanDesc = /^ {2}[A-Za-z0-9 ]+:[A-Fa-f0-9]+$/;
export const isNumber = /^\d+$/;
export const isMac = /^[0-9A-Fa-f]{16}$/;
export const isIPv6 = /^[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+$/;

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function msec2duration(msec: number): number {
  msec = Math.max(msec, 20);
  const duration = Math.floor(Math.log2(msec / 10) - 1);
  return Math.min(duration, 14);
}

export function duration2msec(duration: number): number {
  return Math.pow(2, duration + 1) * 10;
}

export function isValidCache(cache: WsunCache): boolean {
  if (cache.channel < 33 || cache.channel > 60) {
    return false;
  }
  if (cache.panId < 0 || cache.panId > 0xffff) {
    return false;
  }
  if (!isMac.test(cache.addr)) {
    return false;
  }
  return true;
}
