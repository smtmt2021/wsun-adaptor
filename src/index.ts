/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunConfig } from './wsun-interfaces';
import { WsunAdaptor, WsunEventSession } from './wsun-adaptor';
import { WsunBp35a1 } from './wsun-bp35a1';
import DEBUG from 'debug';
const debug = DEBUG('wsun/adaptor');

export { MAX_DGRAM_SIZE } from './wsun-device-constants';
export { WsunAdaptor, WsunEventSession };
export { WsunId, WsunPanDesc, WsunSocket, WsunConfig, WsunCache } from './wsun-interfaces';
export {
  WsunError,
  WsunSendingInProgress,
  WsunUdpTransmissionFailed,
  WsunCommandError,
  WsunTimeoutError
} from './wsun-errors';

export type NameOfClass = 'BP35A1' | 'MOCK';
export interface MockConstructor {
  new (config: WsunConfig): WsunAdaptor;
}

let instance: WsunAdaptor;

/**
 *  Class factory to get a wsun adaptor
 *  @param  className is model name of adaptor you want
 *  @param  config
 *  @return an instance of adaptor which is specified by the model name
 */
export function WsunGetAdaptor(
  className: NameOfClass,
  config?: WsunConfig
): WsunAdaptor | undefined {
  debug('WsunGetAdaptor:', className, JSON.stringify(config));
  if (!instance && config) {
    switch (className) {
      case 'BP35A1':
        instance = new WsunBp35a1(config);
        break;
    }
  }
  return instance;
}
