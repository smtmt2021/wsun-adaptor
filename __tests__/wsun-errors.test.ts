/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunCommandError } from '../src/wsun-errors';

test('WsunCommandError', () => {
  expect(new WsunCommandError(0)).toMatchSnapshot();
  expect(new WsunCommandError(1)).toMatchSnapshot();
  expect(new WsunCommandError(2)).toMatchSnapshot();
  expect(new WsunCommandError(3)).toMatchSnapshot();
  expect(new WsunCommandError(4)).toMatchSnapshot();
  expect(new WsunCommandError(5)).toMatchSnapshot();
  expect(new WsunCommandError(6)).toMatchSnapshot();
  expect(new WsunCommandError(7)).toMatchSnapshot();
  expect(new WsunCommandError(8)).toMatchSnapshot();
  expect(new WsunCommandError(9)).toMatchSnapshot();
  expect(new WsunCommandError(10)).toMatchSnapshot();
  expect(new WsunCommandError(NaN)).toMatchSnapshot();
});
