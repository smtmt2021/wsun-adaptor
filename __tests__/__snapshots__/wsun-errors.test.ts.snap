// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WsunCommandError 1`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 2`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 3`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 4`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 5`] = `[WsunCommandError: wsun: wsun adaptor replied 'unsupported command']`;

exports[`WsunCommandError 6`] = `[WsunCommandError: wsun: wsun adaptor replied 'incorrect command']`;

exports[`WsunCommandError 7`] = `[WsunCommandError: wsun: wsun adaptor replied 'incorrect range']`;

exports[`WsunCommandError 8`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 9`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;

exports[`WsunCommandError 10`] = `[WsunCommandError: wsun: wsun adaptor replied 'uart error']`;

exports[`WsunCommandError 11`] = `[WsunCommandError: wsun: wsun adaptor replied 'failed to get result']`;

exports[`WsunCommandError 12`] = `[WsunCommandError: wsun: wsun adaptor replied 'unknown error']`;
